perf/light.cc

```diff
@@ -0,0 +1,653 @@
+			lookup_iter++;
+			processed += 2;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+// Delete random values till the hashtable is empty.
+BENCHMARK_DEFINE_F(LightFixture, DeleteRandValue)
+(benchmark::State& state)
+{
+	std::int64_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data, state.range(0));
+		data.shuffle(state.range(0));
+		state.ResumeTiming();
+
+		for (auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_delete_value(ht, v.hash, v.raw));
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+BENCHMARK_REGISTER_F(LightFixture, InsertRandValue)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+BENCHMARK_REGISTER_F(LightFixture, FindRandValue)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+BENCHMARK_REGISTER_F(LightFixture, FindRandValueWithMisses)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+BENCHMARK_REGISTER_F(LightFixture, FindRandByKey)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+BENCHMARK_REGISTER_F(LightFixture, SequenceIteration)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+BENCHMARK_REGISTER_F(LightFixture, InsertAfterErase)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+BENCHMARK_REGISTER_F(LightFixture, FindAfterErase)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+BENCHMARK_REGISTER_F(LightFixture, InsertAndFind)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+BENCHMARK_REGISTER_F(LightFixture, DeleteRandValue)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+////////////////////////////// Unordered set Fixture ///////////////////////////////////////////////////////////////////
+
+struct TupleEqual
+{
+	bool operator()(const Tuple &lhs, const Tuple &rhs) const
+	{
+		return std::memcmp(lhs.raw.data, rhs.raw.data, lhs.raw.size) == 0;
+	}
+};
+
+// Just to compare with Light. These measurements are not assumed to be the
+// most closest to ones related to Light - hash tables interfaces seem to be
+// too different. It's just about making Light numbers a bit representative.
+class UnorderedSetFixture : public benchmark::Fixture {
+public:
+	void
+	TearDown(const benchmark::State &state)
+	{
+		set.clear();
+	}
+
+protected:
+	void
+	Fill(TupleHolder &holder, std::size_t count)
+	{
+		assert(count <= holder.tuples.size());
+		set.insert(holder.tuples.begin(), holder.tuples.begin() + count);
+	}
+
+	void
+	Erase(TupleHolder &holder, std::size_t count)
+	{
+		assert(count <= holder.tuples.size());
+		for (auto itr = holder.tuples.begin(); itr < holder.tuples.begin() + count; ++itr) {
+			set.erase(*itr);
```

Comment:
Please use range-based `erase` overload.

Reply:
Agree. Usual practice: shorter code - less questions.

Reply:
@alyapunov please clarify which of approaches in more general (I guess the one with for_each but to newbie like me it's not clear).

Reply:
Fixed, now Fill and Erase take iterators.